Cloud init schema failures

Chris-Peterson444 opened the pull request:
Users attempting to do autoinstall may incorrectly send autoinstall directives as cloud-config, which will result in cloud-init schema validation errors. When loading autoinstall from cloud-config, we now check to see if there are any cloud-init schema validation errors and warn the user. Additionally, if the source of the error is from a known autoinstall error, we inform the user and halt the installation with a nonreportable AutoinstallError.

Requires #1945 and #1947.

dbungert commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +69,56 @@ def read_legacy_status(stream):
+    cmd: list[str] = ["cloud-init", "schema", "--system"]
+    status_coro: Awaitable = arun_command(cmd)
+    try:
+        sp: CompletedProcess = await asyncio.wait_for(status_coro, 600)
```
So while the big wait on `cloud-init status --wait` is allowed to be 10 minutes, most of the others should be instantaneous.  Can you do some real vm testing on the typical response times for these calls?  I'm thinking `timeout=10` might be more realistic.  Part of that 10 minutes is the expectation that the system may still be starting, cloud-init is fetching the data from the datasource, etc.  Whereas the calls here should be much quicker.

Chris-Peterson444 replied:
Good point. In my tests they seem pretty instantaneous, even with a large config with many incorrect keys. I've updated the timeouts to the suggested 10 seconds.

dbungert commented on subiquity/cloudinit.py:
```diff
@@ -86,3 +138,31 @@ async def cloud_init_status_wait() -> (bool, Optional[str]):
+    cmd: list[str] = ["cloud-init", "status", "--format=json"]
+    status_coro: Awaitable = arun_command(cmd)
+    try:
+        sp: CompletedProcess = await asyncio.wait_for(status_coro, 600)
```
see above

dbungert reviewed (commented):
A comment for now, more review tomorrow.

Chris-Peterson444 commented:
@blackboxsw could you please take a look at this as well?

dbungert commented on subiquity/cloudinit.py:
```diff
@@ -86,3 +138,31 @@ async def cloud_init_status_wait() -> (bool, Optional[str]):
+    """
+
+    if not supports_recoverable_errors():
+        return (True, None)
```
I guess it would be more pythonic to have this raise with an exception with the list of bad keys, which could in turn be re-raised as that `AutoinstallError`.  What do you think?

Chris-Peterson444 replied:
Yeah I think that makes sense. I'll rework the logic. 

Since that if section is highlighted: Part of my assumption here was that we needed recoverable error support to get the bad keys. i.e. we need to ask cloud-init if invalid cloud-config was something it encountered. I suppose this isn't necessarily true, and we could just query the schema always. This has the benefit of cutting out one of the subprocess calls. What do you think?

dbungert replied:
That makes sense to me but let's VM confirm it with older images, where `supports_recoverable_errors() == False`

dbungert commented on subiquity/server/server.py:
```diff
@@ -723,7 +727,66 @@ async def start_site(self, runner: web.AppRunner):
+        context.set("is-install-context", True)
+        context.enter()  # publish start event
+
+        if "autoinstall" not in cloud_cfg:
```
Well, part of the problem is that people aren't supplying an autoinstall key, but supplying autoinstall directives directly to cloud-init  (or could be doing both, but I'm guessing that's not common).  So I think we need to revisit this part.  retrieving `cloud_cfg["autoinstall"]` could be after `validate_cloud_init_schema()` in the current setup.

Chris-Peterson444 replied:
That makes sense to me. This probably flows better with the throws-on-failure logic.

dbungert reviewed (commented):
Chris, I'm really glad to see this work done.  I think it's going to help reduce user confusion.  I have some items to follow-up on.

Chris-Peterson444 commented:
Thanks for the review @dbungert. I've gone ahead and implemented your suggestions. I added an extra test case to capture the scenario in which autoinstall keys are present in the cloud-config and the `autoinstall` key itself is not.

Also, by removing the dependence on the recoverable errors I was able to test this successfully on a 20.04.6 image which has cloud-init version `22.4.2-0ubuntu0~20.04.2`

ogayot commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +85,56 @@ def read_legacy_status(stream):
+    try:
+        sp: CompletedProcess = await asyncio.wait_for(status_coro, 10)
+    except asyncio.TimeoutError:
+        return []
```
ditto, a warning would be good I think :)

ogayot commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +85,56 @@ def read_legacy_status(stream):
+    """Retrieve the keys causing schema failure."""
+
+    cmd: list[str] = ["cloud-init", "schema", "--system"]
+    status_coro: Awaitable = arun_command(cmd)
```
I think it's worth passing `clean_locale=True` explicitly (even though it's the default) since we're potentially parsing locale dependent output.

ogayot commented on subiquity/cloudinit.py:
```diff
@@ -86,3 +154,23 @@ async def cloud_init_status_wait() -> (bool, Optional[str]):
     return (True, status)
+
+
+async def validate_cloud_init_schema() -> bool:
```
Maybe make the function return None since we're going to ignore the return value anyway?

ogayot commented on subiquity/cloudinit.py:
```diff
@@ -86,3 +154,23 @@ async def cloud_init_status_wait() -> (bool, Optional[str]):
+    if causes == []:
+        return True
+    else:
+        raise CloudInitSchemaValidationError(keys=causes)
```
suggestion:

```python
causes: list[str] = await get_schema_failure_sources()

if causes:
    raise CloudInitSchemaValidationError(keys=causes)
```

ogayot commented on subiquity/server/tests/test_server.py:
```diff
@@ -357,6 +359,92 @@ async def test_autoinstall_validation__filter_autoinstall(self, config, good, ba
+                val_mock.return_value = True
+            else:
+                val_mock.side_effect = CloudInitSchemaValidationError(
+                    keys=[key for key in cloud_data.keys()]
```
suggestion

keys=list(cloud_data.keys())

ogayot commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +85,56 @@ def read_legacy_status(stream):
+    try:
+        status: dict[str, Any] = json.loads(stream)
+    except json.JSONDecodeError:
+        return []
```
I think a warning would be good here. Not sure if we should try to display the contents of `stream` in the warning message though...

Chris-Peterson444 replied:
I think that's a good idea. It has to expect the output from `cloud-init status --format=json` so I think something like "Failed to decode cloud-init status output" is informative enough.

ogayot reviewed (approved):
Good work! A few minor comments but nothing blocking

Chris-Peterson444 commented:
Thanks @ogayot I've implemented all of your suggestions

dbungert reviewed (approved):
One tweak please then LGTM.

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -4,13 +4,31 @@
 log = logging.getLogger("subiquity.cloudinit")
 
 
+class CloudInitSchemaValidationError(Exception):
```
I think that's reasonable, let's do that. I'll update shortly.

Chris-Peterson444 commented:
Thanks! Since I changed CloudInitSchemaValidationError to a NonReportableException, I also added some lines to the client code so the error overlay works for those error types. Could you give that a look over too @dbungert ?

blackboxsw reviewed (commented):
Thanks for the ping on this for our awareness.

While I think consuming the CLI and reading stderr is potentially fragile, I get that subiquity may not want to tightly couple to internal cloudinit python libraries and functions to process and iterate over SchemaValidationErrors 
raised by [cloudinit.config.schema.validate_cloudconfig_schema](
https://github.com/canonical/cloud-init/blob/main/cloudinit/config/schema.py#L685) or [cloudnit.config.schema.validate_cloudconfig_file)[https://github.com/canonical/cloud-init/blob/main/cloudinit/config/schema.py#L1037]. 

To use these functions, subiquity would have to likely have to do something like the following:
```
from cloudinit.config.schema import validate_cloudconfig_file, get_schema, SchemaValidationError
cloudconfig_schema = get_schema()
try:
    validate_cloudconfig_file('/var/lib/cloud/instance/user-data.txt', schema=cloudconfig_schema)
except SchemaValidationError as e:
    return [schema_problem.path for schema_problem in e.schema_errors if "unexpected" in schema_problem.message]
```
Note that even the structured SchemaValidationError will not set a 'path' attribute if multiple unexpected keys exist on the object. So, you'd still need your pattern matching to `schema_problem.message` to extract the separate keys so it's nearly the same logic you have for parsing the stderr of `cloud-init schema --system`.

The fragility in CLI approach in this PR will be due to cloud-init relying on `jsonschema` for that error string as cloud-init just presents that error message directly without modification. If `jsonschema` module changes their error messaging format this could break subiquity parsing.

That said, I don't see this error output format from jsonschema  being any different between jammy and noble and I think cloud-init would like to work on a machine-readable representation of `cloud-init schema --system --format=yaml` per your feature request https://github.com/canonical/cloud-init/issues/5100 that will make this easier to process in the future.

Minor changes requested that you can take or leave as you see fit
- better regex pattern match, splitting of the parsed jsonschema error messages
- dropping unused functions/tests
- 
We'll make sure we keep you informed when we start tackling https://github.com/canonical/cloud-init/issues/5100. So this code can consume more friendly structured content when available.

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +86,58 @@ def read_legacy_status(stream):
     return None
 
 
+def read_recoverable_errors(
```
Yeah this was part of an earlier implementation. Sorry about that. I thought to keep it around but I'll be removing it for now since it's not in use.

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +86,58 @@ def read_legacy_status(stream):
+    """Retrieve recoverable errors from cloud-init status."""
+
+    try:
+        status: dict[str, Any] = json.loads(stream)
```
As above, dropped for now. Thanks!

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +85,58 @@ def read_legacy_status(stream):
+    if level is not None:
+        return errors_by_level.get(level, [])
+
+    all_errors: list[str] = []
+    for _level in errors_by_level:
+        all_errors.extend(errors_by_level[_level])
+
+    return all_errors
```
Ah TIL. Thanks for the recommendation

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +86,58 @@ def read_legacy_status(stream):
+    return all_errors
+
+
+async def get_schema_failure_sources() -> list[str]:
```
Agreed, changing. Thanks

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +86,58 @@ def read_legacy_status(stream):
+async def get_schema_failure_sources() -> list[str]:
+    """Retrieve the keys causing schema failure."""
+
+    cmd: list[str] = ["cloud-init", "schema", "--system"]
```
I think in general we want to capture errors from all sources since we read the combined config when extracting the autoinstall config (e.g. is it possible someone sends autoinstall config in their network-config?), but capturing the file source would improve our error messaging for sure. What do you think about leaving this to future improvements?

Chris-Peterson444 commented on subiquity/cloudinit.py:
```diff
@@ -67,6 +86,58 @@ def read_legacy_status(stream):
+    # Matches:
+    # ('some-key' was unexpected)
+    # ('some-key', 'another-key' were unexpected)
+    pattern = r"\((?P<args>'[\w\-]+'(,\s'[\w\-]+')*)+ (?:was|were) unexpected\)"
```
Thanks for the review on this espeically (Regex is hard!). I think we can definitely (1) replace `[\w\-]` with `[^']` to be more flexible and (3) remove the trailing `+`, but (2) isolating the key names without the quotes is difficult. The suggested regex doesn't quite work and my attempts thus far have been insufficient.

I think I'm okay with another round of processing to strip the quotes and this is still an improvement on the parsing.

Chris-Peterson444 commented on subiquity/tests/test_cloudinit.py:
```diff
@@ -132,3 +137,114 @@ async def test_cloud_init_status_wait_legacy(self, m_wait_for):
         self.assertEqual((True, "done"), await cloud_init_status_wait())
+
+
+class TestCloudInitUtil(SubiTestCase):
```
As above, correct. removing!

Chris-Peterson444 commented:
@blackboxsw Thanks a lot for the extensive review! I went ahead and implemented most of your suggested changes (all but two, I left my reasoning in the comments and kept them unresolved). Based on your analysis I think we're okay with the potential fragility in the error parsing until we move to the structured output when it's available. I look forward to seeing more on https://github.com/canonical/cloud-init/issues/5100!

dbungert commented:
OK to merge, despite Noble test failures (unrelated archive problems)
